File: apps/apple/Sources/SupportExportCore/SupportApplicationSource.swift
import Foundation

/// Immutable application build metadata; scanner provenance cannot supply a missing application source.
public struct SupportApplicationSource: Encodable, Sendable {
    public let producer = "application-build"
    public let observation = "current"
    public let sourceSha: String
    public let treeSha: String

    /// Unstamped local builds return nil; partial, unresolved or malformed metadata refuses export.
    public init?(info: [String: Any]) throws {
        let source = info["DSHApplicationSourceSHA"]
        let tree = info["DSHApplicationSourceTree"]
        if source == nil && tree == nil { return nil }
        guard let sourceSha = source as? String, let treeSha = tree as? String else {
            throw SupportExportError.invalidIdentity
        }
        if sourceSha.isEmpty && treeSha.isEmpty { return nil }
        for value in [sourceSha, treeSha] {
            guard value.utf8.count == 40, value.utf8.allSatisfy({
                (48...57).contains($0) || (97...102).contains($0)
            }) else { throw SupportExportError.invalidIdentity }
        }
        self.sourceSha = sourceSha
        self.treeSha = treeSha
    }
}

File: apps/android/core/src/main/kotlin/ai/deepseek/dsh/companion/SupportApplicationSource.kt
package ai.deepseek.dsh.companion

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** Application commit and tree are independent of the scanner library's build metadata. */
data class SupportApplicationSource(val sourceSha: String, val treeSha: String) {
    init {
        if (listOf(sourceSha, treeSha).any { !Regex("[a-f0-9]{40}").matches(it) }) {
            throw SupportExportException(SupportExportFailure.INVALID_IDENTITY)
        }
    }

    companion object {
        /** Unstamped builds have neither field; partial, mistyped or unresolved metadata is invalid. */
        fun fromMetadata(source: Any?, tree: Any?): SupportApplicationSource? {
            if (source == null && tree == null || source == "" && tree == "") return null
            if (source !is String || tree !is String) throw SupportExportException(SupportExportFailure.INVALID_IDENTITY)
            return SupportApplicationSource(source, tree)
        }
    }

    internal fun toJson() = buildJsonObject {
        put("producer", "application-build")
        put("observation", "current")
        put("sourceSha", sourceSha)
        put("treeSha", treeSha)
    }
}
